### DISCUSSION TITLE
Extraction of pore network using PoreSpy

### DISCUSSION THREAD

**Discussion by Soheil1993** [PMEAL/porespy#679] [Q&A]
Hello;

I am Soheil Safari, a first-year PhD student in Hydrogeology at Charles University in Prague.

A part of my PhD project is investigation of the effects of compaction on pore structure and hydraulic properties in granular porous media. 
For sphere packing and compaction I used YADE-DEM software (with Python interface), but I could not extract the pore network with it. I attached the result for your consideration.

Would you please let me is it possible to extract the pore network (pore body and pore throat) with PoreSpy and use it for pore network modeling with openPNM?

How can I import my results to PoreSpy? In which format should I save them?

I am grateful for any hints or insights into this issue.

Thank you so much in advance, and I look forward to hearing from you.

Best regards,
Soheil Safari

![my result](https://user-images.githubusercontent.com/94166206/161780403-32ac6446-33aa-4d03-b9a8-3429fadd8c34.png)

**Comment by ma-sadeghi** [comment]
Hi @Soheil1993, for extracting a pore network using `porespy`, what you need is a voxel image (a 3d array containing 0s and 1s, which refer to void and solid, respectively). Probably, you can get such an image directly from YADE-DEM. Maybe look for "Export to TIFF" or something?

**Comment by Soheil1993** [comment]
Dear Dr. @ma-sadeghi 

So sorry for the late response.

Thank you very much for your kind reply and valuable information.

I can save the images of YADE in (.vtk) format for post-processing in Paraview. Is it also proper for using in porespy?

Would you please let me know how I can import it to the porespy?
Can I use this function which you recommended to me before?

```python
import imageio
im = imageio.volread(path_to_your_image)  # for 3D images
```

How can I extract the pore network with porespy (Both pore and throat) and export to the openPNM?

I should mention that I'm a bit new to Python.

Thank you so much in advance, and I look forward to hearing from you.

Best regards,
Soheil

**Reply by jgostick** [reply]
Hi @Soheil1993.  You can find a tutorial for extracting networks here: https://porespy.org/examples.html#networks

**Reply by Soheil1993** [reply]
Dear Dr. @jgostick 
Thank you very much for your kind reply and the great link you shared with me.

**Reply by ma-sadeghi** [reply]
@Soheil1993 You should be able to turn VTK into a numpy array using the `VTK` library. You can install it using `conda install vtk`. I'm sure you can find what you find with a little bit of Googling. [Here](https://discourse.vtk.org/t/convert-vtk-array-to-numpy-array/3152/2)'s one that I found.

**Reply by jgostick** [reply]
I am sorry, but I think I confused your question with another similar one.  Let me chip in an answer here.  Amin's suggestion is an option, but you could also create a blank image of 0's and insert spheres of known radii at given center points, assuming you can get this from Yade.  I have given a bit of instruction [here](https://github.com/PMEAL/porespy/discussions/678#discussioncomment-2502897).  The difference in your case is that you need to insert spheres of different size.  This is also possible, but instead of inserting a single pixel at each center point, you can use the ``porespy.tools.insert_sphere`` method, which accepts a center point AND a radius.  It's not fast at the moment, but does work.

**Reply by jgostick** [reply]
I have posted a function for generating an image if the sphere centers and their radii are known, here: #680

**Reply by Soheil1993** [reply]
Dear Dr. @jgostick and Dr. @ma-sadeghi 

Many thanks for your valuable information and useful links.

I will go through them. I hope I can get proper result.

Best regards,
Soheil
